### ISSUE TITLE
Fine-grained macaroon for external services

### ISSUE THREAD

**Issue by Zetanova** [lightningnetwork/lnd#4510]
## Intro
After the first use of lnd, many node owners discover there need for more common features
like channel-management, backup, monitoring, rebalance and more.

There are view options already out that are fulfilling this common features,
but all of them require self-hosting of there components 
and most of the time an admin-macaroon.

## Feature Request
The macaroon system could be fine grained to combine them to support multiple external-service providers
and extended with an 2FA system to request permissions from the node operator.
This could be implemented directly in lnd or as a reverse-proxy 
that has its own macaroon system to limit the features.

## Examples
*) backup service
somebody is hosting a public or private external backup service
that is subscribing to the gRPC backup-events or pulling the backup in intervals.
If the provider of this service is not the same as the node-owner
it does not require a full-readonly-macaroon to the node
https://github.com/lightningnetwork/lnd/issues/4448

*) monitoring service
somebody is hosting a public or private external healthcheck service
A node owner could share it's healthcheck service feature with other node operators.
The healthcheck-operator could request an inbound channel as "payment"
and could notify the node-owner on problems and even unlock the node on user request
when it's restarted  with a unlock-only macaroon or  provide an 
[Tang Server](https://github.com/latchset/tang)

*) rebalancing Interface
Loop could provide users it's [lightning terminal](https://github.com/lightninglabs/lightning-terminal) directly. 
It could accept ln-founds or just an inbound channel as payment.
It would need beside a readonly-macaroon an 2FA-macaroon to execute loop-out transfers or channel-closures.

*) identity service
An external service that acts as an agent for the lnd node could request the lnd-node to sign a message 
to use as a verification on an webite like https://1ml.com/ (and others) and subscribe and fill out the node profile.

**Comment by Roasbeef** [comment]
I think https://github.com/lightningnetwork/lnd/pull/4463 makes most of those referenced use cases possible?

**Comment by Zetanova** [comment]
@Roasbeef  #4463 looks good for it, but i don't see how it will perform with a macaron that includes half of the rpc-methods.  
Maybe  it can be resolved with somekind of custom-permission that groups uri permissions together, 
like `perm:myGroup` (includes uri:/lnrpc.Lightning/GetInfo, uri:...)

The feature for **2FA** to allow the node-owner to permit an execution of a critical process 
would be essential for many features too. I don't know how it can be archived with only macaroons. 
Beside the macaroon permission system, 
lnd need to have somekind of possibility of request and sign command-messages.
The flow would be:
1) External-Service request command-XYZ to be signed
2) Node-owner gets a permission-notification (phone or somehting else, human readable) and signs it. 
2) LND calls back the external-service with the signed-command-message as payload
3) External-Service execute/sends the signed-command to the service (back to LND or maybe Loop, ...)

Basically a service with an admin-macaroon need an option to verify a signed message from the node/owner
to allow not only the gRPC method execution but the execution of a single command-message. 

The option with an optional/standalone **ln-grpc-reverse-proxy** that would act as an **gateway**
for customized macaroon and **2FA authentication** should not be dismissed.

**Comment by Zetanova** [comment]
I already have an proposal for the signed message problem.
The permission system could be extened with a uri-fragment `#signed` 

A macaroon could then include:
```
uri:/lnrpc.Lightning/GetInfo
invoices:read
invoices:write#signed
```
This would mark all grpc-operators that require the "invoices:write" permission to also have 
a valid signatur of the command-message.

**Comment by Roasbeef** [comment]
Possible now with the macaroon bakery.
